File: python/tests/test_hooke.py
import numpy as np

from numpy.testing import assert_array_equal

from pyhd98 import Hooke


def test_hooke_data():
    lambda_ = 1.2
    mu = 3.4
    hooke = Hooke(lambda_, mu)
    assert hooke.lambda_ == lambda_
    assert hooke.mu == mu


def test_hooke_current_state():
    lambda_ = 1.2
    mu = 3.4
    hooke = Hooke(lambda_, mu)

    dim = 3
    sym = 6
    I = np.zeros((sym,), dtype=np.float64)
    I[:dim] = 1.0

    eps = np.array([1.2, -3.4, 5.6, -7.8, 9.1, -2.3], dtype=np.float64)
    sig_act = np.zeros_like(I)
    hooke.current_state(eps, np.array(0.0), sig_act)

    sig_exp = np.zeros_like(I)
    tr_eps = eps[:dim].sum()
    sig_exp = lambda_ * tr_eps * I + 2 * mu * eps

    assert_array_equal(sig_act, sig_exp)


def test_hooke_update():
    lambda_ = 1.2
    mu = 3.4
    hooke = Hooke(lambda_, mu)

    dim = 3
    sym = 6
    I = np.zeros((sym,), dtype=np.float64)
    I[:dim] = 1.0

    eps1 = np.zeros_like(I)
    delta_eps = np.array([1.2, -3.4, 5.6, -7.8, 9.1, -2.3], dtype=np.float64)

    sig2_act = np.zeros_like(I)
    C2_act = np.zeros((sym, sym), dtype=np.float64)
    hooke.update(delta_eps, eps1, np.array(0.0), sig2_act, np.array(0.0), C2_act)

    tr_eps = delta_eps[:dim].sum()
    sig2_exp = lambda_ * tr_eps * I + 2 * mu * delta_eps

    C2_exp = lambda_ * I[:, None] * I[None, :] + 2 * mu * np.eye(sym)

    assert_array_equal(sig2_act, sig2_exp)
    assert_array_equal(C2_act, C2_exp)

File: tests/test_hooke.cpp
#include <array>
#include <iostream>

#include <catch2/catch.hpp>

#include "hd98/hooke.hpp"
#include "test_hd98.hpp"

static void test_current_state(hd98::Hooke const& mat) {
  Tensor2 eps{};
  Tensor2 sig_act{};
  Tensor2 sig_exp{};
  for (size_t i = 0; i < hd98::sym; i++) {
    eps[i] = 1.;
    mat.current_state(eps.data(), nullptr, sig_act.data());
    std::fill(sig_exp.begin(), sig_exp.end(), 0.0);
    sig_exp[i] = 2 * mat.mu * eps[i];
    if (i < hd98::dim) {
      for (size_t j = 0; j < hd98::dim; j++) {
        sig_exp[j] += mat.lambda;
      }
    }
    assert_array_equal(hd98::sym, sig_act.data(), sig_exp.data(), 1e-15, 1e-15);
    eps[i] = 0.;
  }
}

static void test_update(hd98::Hooke const& mat) {
  Tensor2 eps1{};
  Tensor2 delta_eps{};
  Tensor2 sig2_act{};
  Tensor2 sig2_exp{};
  Tensor4 C2_act{};
  Tensor4 C2_exp{};
  for (size_t i = 0, ij = 0; i < hd98::sym; i++) {
    for (size_t j = 0; j < hd98::sym; j++, ij++) {
      C2_exp[ij] = (i < hd98::dim) && (j < hd98::dim) ? mat.lambda : 0.;
      if (i == j) {
        C2_exp[ij] += 2 * mat.mu;
      }
    }
  }
  for (size_t i = 0; i < hd98::sym; i++) {
    delta_eps[i] = 1.;
    mat.update(delta_eps.data(), eps1.data(), nullptr, sig2_act.data(), nullptr,
               C2_act.data());
    mat.current_state(delta_eps.data(), nullptr, sig2_exp.data());
    assert_array_equal(hd98::sym, sig2_act.data(), sig2_exp.data(), 1e-15,
                       1e-15);
    assert_array_equal(hd98::sym * hd98::sym, C2_act.data(), C2_exp.data(),
                       1e-15, 1e-15);
    delta_eps[i] = 0.;
  }
}

TEST_CASE("Hooke") {
  double const mu = 1.2;
  double const nu = 0.3;
  double const lambda = 2 * mu * nu / (1 - 2 * nu);
  hd98::Hooke const mat{lambda, mu};

  SECTION("current_state") { test_current_state(mat); }
  SECTION("Update") { test_update(mat); }
}